### DISCUSSION TITLE
using the engine with multiple processes?

### DISCUSSION THREAD

**Discussion by nishantvarma** [sqlalchemy/sqlalchemy#11583] [Usage Questions]
**Context**

We have a `Zope` application that mixes `session.execute`, `session.query` and `legacy_query` like this:
```
def request_api(): # web request
    result = session.query(Table).all() # orm
    result = session.execute(query).fetchall() # sql
    result = legacy_query(query) # legacy
    ...
```

`legacy_query` is a modified version of [Products.SQLAlchemyDA.query](https://github.com/zopefoundation/Products.SQLAlchemyDA/blob/master/src/Products/SQLAlchemyDA/da.py#L352-L412). The patching was done to use ORM, and re-use the same session in (legacy) ZSQL's.
```
def legacy_query(query):
    connection = Session().connection().connection # scoped-session.
    cursor = connection.cursor()
    cursor.execute(query)
    description = cursor.description
    if description is not None:
        rows = cursor.fetchall()
    return rows
```
**Problem**

Under heavy load we see some "Connection is busy" errors. It is reported from **multiple processes** in the **same machine** at the **same time** (`zope_c1`, `zope_c2` etc. with 4 threads each):
```
zope_c1.log
-----------
[waitress 1]
10:00 AM: Connection is busy

zope_c2.log
-----------
[waitress 1]
10:00 AM: Connection is busy
```
**Solution Attempt**

Assuming this error occurs if we use **multiple cursors** from the **same connection** without consuming the previous result (`fetchall` or `close`),  I tried [closing](https://github.com/mkleehammer/pyodbc/issues/1366) the cursor using `closing` context manager, but that doesn't seem to help (perhaps reduced?). I got this hint from an old [post](https://groups.google.com/g/sqlalchemy/c/5SqWfF1QL_Q) from @zzzeek. I was suspecting GC delay (ref count of local varaible `cursor` was 2) under heavy load. However, that doesn't explain why **multiple processes** could have the **same issue** at the **same time**.

1. Does `session.connection().connection` always the give the connection associated to `Session()`? Hope there are no connection mixups happening here. Wonder if any lock is required in this case? 
2. Is there a way to close previous cursors before executing the current query?
3. Any retry mechanisms?
4. Is MARS really required in our case? We don't really fire **multiple queries (cursors)** from the **same connection** unless there is some lazy load happening under the hood. We always use `session.query().all` and `session.execute.fetchall()`.

Any thoughts in debugging would be helpful as well.

PS: We use SQLAlchemy 1.3 in this case; we are in the process of migrating to 1.4 and then 2.x.

**Comment by zzzeek** [comment]
> Does session.connection().connection always the give the connection associated to Session()? Hope there are no connection mixups happening here. 

yes

> Wonder if any lock is required in this case?

probably not, the program should be fixed so that there is no concurrent use of a single Connection.  use separate connections/sessions per thread if there's concurrency in place here.    If the behavior is local to a single thread and the program just is leaving a cursor open while it runs another query, that should just be fixed, or a DBAPI driver that supports multiple simultaneous cursors (if one is available, this may not be the case) should be used.

> Is there a way to close previous cursors before executing the current query?

per https://docs.sqlalchemy.org/en/20/core/connections.html#basic-usage 

> The object returned is known as [CursorResult](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.CursorResult), which references a DBAPI cursor and provides methods for fetching rows similar to that of the DBAPI cursor. The DBAPI cursor will be closed by the [CursorResult](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.CursorResult) when all of its result rows (if any) are exhausted. A [CursorResult](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.CursorResult) that returns no rows, such as that of an UPDATE statement (without any returned rows), releases cursor resources immediately upon construction.

> Any retry mechanisms?

Retrying on a locked cursor would be ill-advised here, as this is not a typical unavoidable "retry" scenario , connections should be used correctly in the first place, that is, if the driver supports only single cursor use (such as MySQL drivers), cursors should be exhausted before running a new query.


> Is MARS really required in our case? We don't really fire multiple queries (cursors) from the same connection unless there is some lazy load happening under the hood. We always use session.query().all and session.execute.fetchall().

"MARS", googling this unfamiliar term seems to indicate you're using SQL Server.    If the lack of MARS is why you are getting these "connection is busy" log messages and turning it on makes them go away, maybe?      

Since the issue occurs with multiple processes and I now know you're using SQL Server, maybe you have some ODBC settings like connection pooling that you can look at as well.

> unless there is some lazy load happening under the hood.

When you get results from the ORM, assuming you are not using yield_per, the database result object is already fully fetched and the cursor is closed.  lazy loads while you iterate over a database result from the ORM would not present a problem.

**Reply by zzzeek** [reply]
oh, multiple processes, sorry, one more important thing that is likely your issue:

**how are you spawning these processes and sharing your database engine across processes?**

You want to ensure **no connection pools are shared across processes**.

Follow the guidance at https://docs.sqlalchemy.org/en/20/core/pooling.html#using-connection-pools-with-multiprocessing-or-os-fork

**Comment by nishantvarma** [comment]
Thanks for the pointers, @zzzeek. We are seeing that Zope 5 + Waitress is sharing the request object (and hence connection) across multiple threads:

```
2024-07-10 10:11:17,896 INFO [SQLAlchemyDA:96][waitress-1] Zope Request Id 140472489357520, DB session Id 140472489360064 DB connection id 140472380988176
2024-07-10 10:11:17,384 INFO [SQLAlchemyDA:96][waitress-2] Zope Request Id 140472489357520, DB Session Id 140472489360064 DB connection Id 140472380988176
```

It leads to error when multiple threads fire query (there is a gap in execute and fetchall). Will debug further. Will close the discussion for now. Appreciate your help.

Just as an aside: isn't it better to use the `closing` decorator when using `cursor` directly, especially if an exception occurs - or is that automatically taken care? A lot of ORM code does that (safe_close_cursor stuff).
